File: python/703.py
"""Project Euler Problem 703: Circular Logic III.

Let f be a function from N-tuples of booleans (b_1, ... b_N) to N-tuples of
booleans (c_1, ... c_N) where c_i = b_{i+1} for i<N and c_N = b_1 AND
(b_2 XOR b_3). Find the number of functions T from N-tuples of booleans to
booleans such that T(x) AND T(f(x)) = false for all N-tuples x.

We can construct a directed graph with edges from each x to f(x). Each vertex
has a single outgoing edge, so each connected component is a tree with a
single additional edge from the root to some vertex. The problem is then to
find the number of ways to color each vertex of this graph either false (f)
or true (t) such that no two true vertices are connected.

Each connected component is independent. For each connected component, the
root can be either false or true. If it's false, then we can ignore the
constraint on the additional edge, and we can recurse on the tree. Let f_v
and t_v be the number of colorings of a subtree rooted at v if the root v is
colored false and true, respectively. Then we can compute f_v and t_v by
computing f_u and t_u for all children u; we have f_v = Π_u (f_u + t_u) and
t_v = Π_u f_u.

If the root is true, then the vertex at the other end of the additional edge
(call it n) must be false. So we can cut the tree at n, and compute the
number of colorings of the two subtree pieces.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List


@dataclass
class Result:
    """Result for subtree coloring."""

    f: int  # Number of colorings with root false
    t: int  # Number of colorings with root true


def solve() -> int:
    """Solve Problem 703."""
    n = 20
    m = 1_001_001_011

    # Build next array: next[i] = f(i)
    next_arr: List[int] = []
    for i in range(1 << n):
        # c_i = b_{i+1} for i < N
        # c_N = b_1 AND (b_2 XOR b_3)
        # So: shift right by 1, then set highest bit to b_1 AND (b_2 XOR b_3)
        shifted = i >> 1
        b1 = i & 1
        b2 = (i >> 1) & 1
        b3 = (i >> 2) & 1
        highest_bit = b1 & (b2 ^ b3)
        next_val = shifted + (highest_bit << (n - 1))
        next_arr.append(next_val)

    # Build reverse graph (prevs[i] = list of nodes that map to i)
    prevs: List[List[int]] = [[] for _ in range(1 << n)]
    for i in range(1 << n):
        prevs[next_arr[i]].append(i)

    used = [False] * (1 << n)
    ans = 1

    def helper(ptr: int, avoid: int) -> Result:
        """Compute colorings for subtree rooted at ptr, avoiding avoid."""
        used[ptr] = True
        if ptr == avoid:
            return Result(1, 0)

        f_val = 1
        t_val = 1
        for prev in prevs[ptr]:
            result = helper(prev, avoid)
            f_val = (f_val * (result.f + result.t)) % m
            t_val = (t_val * result.f) % m

        return Result(f_val % m, t_val % m)

    for i in range(1 << n):
        if not used[i]:
            # Find root of cycle
            root = i
            while not used[root]:
                used[root] = True
                root = next_arr[root]

            # Root is false: all subtrees independent
            f_val = 1
            for prev in prevs[root]:
                res = helper(prev, root)
                f_val = (f_val * (res.f + res.t)) % m

            # Root is true: next[root] must be false
            if next_arr[root] == root:
                t_val = 0
            else:
                t_val = (
                    helper(next_arr[root], root).f
                    * helper(root, next_arr[root]).t
                ) % m

            ans = (ans * (f_val + t_val)) % m

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/703.c
/*
 * Project Euler Problem 703: Circular Logic III.
 *
 * Build a functional graph from f on N-bit boolean tuples, where f shifts
 * right and sets the high bit to b1 AND (b2 XOR b3). Count independent set
 * colorings on each connected component.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NN 20
#define SZ (1 << NN)
#define MOD 1001001011LL

typedef long long ll;

int next_arr[SZ];
int *prevs[SZ];
int prev_count[SZ];
int prev_cap[SZ];
int used[SZ];

typedef struct {
    ll f, t;
} Result;

/* Stack-based DFS to avoid recursion overflow */
typedef struct {
    int ptr;
    int avoid;
    int child_idx;
    ll f_val, t_val;
} Frame;

Frame stack[SZ + 10];

Result helper_iterative(int start_ptr, int start_avoid) {
    int sp = 0;
    stack[0].ptr = start_ptr;
    stack[0].avoid = start_avoid;
    stack[0].child_idx = -1;
    stack[0].f_val = 1;
    stack[0].t_val = 1;

    while (sp >= 0) {
        Frame *fr = &stack[sp];
        int ptr = fr->ptr;

        if (fr->child_idx == -1) {
            used[ptr] = 1;
            if (ptr == fr->avoid) {
                /* Return Result(1, 0) */
                if (sp == 0) {
                    return (Result){1, 0};
                }
                sp--;
                Frame *parent = &stack[sp];
                Result child_res = {1, 0};
                parent->f_val = (parent->f_val * ((child_res.f + child_res.t) % MOD)) % MOD;
                parent->t_val = (parent->t_val * child_res.f) % MOD;
                parent->child_idx++;
                continue;
            }
            fr->child_idx = 0;
        }

        if (fr->child_idx < prev_count[ptr]) {
            int child = prevs[ptr][fr->child_idx];
            /* Push child */
            sp++;
            stack[sp].ptr = child;
            stack[sp].avoid = fr->avoid;
            stack[sp].child_idx = -1;
            stack[sp].f_val = 1;
            stack[sp].t_val = 1;
            continue;
        }

        /* All children processed */
        Result res = {fr->f_val % MOD, fr->t_val % MOD};
        if (sp == 0) {
            return res;
        }
        sp--;
        Frame *parent = &stack[sp];
        parent->f_val = (parent->f_val * ((res.f + res.t) % MOD)) % MOD;
        parent->t_val = (parent->t_val * res.f) % MOD;
        parent->child_idx++;
    }

    return (Result){1, 1}; /* Should not reach here */
}

void add_prev(int node, int prev) {
    if (prev_count[node] >= prev_cap[node]) {
        prev_cap[node] = prev_cap[node] ? prev_cap[node] * 2 : 4;
        prevs[node] = realloc(prevs[node], prev_cap[node] * sizeof(int));
    }
    prevs[node][prev_count[node]++] = prev;
}

int main() {
    /* Build next array */
    for (int i = 0; i < SZ; i++) {
        int shifted = i >> 1;
        int b1 = i & 1;
        int b2 = (i >> 1) & 1;
        int b3 = (i >> 2) & 1;
        int highest_bit = b1 & (b2 ^ b3);
        next_arr[i] = shifted + (highest_bit << (NN - 1));
    }

    /* Build reverse graph */
    memset(prev_count, 0, sizeof(prev_count));
    memset(prev_cap, 0, sizeof(prev_cap));
    memset(prevs, 0, sizeof(prevs));
    for (int i = 0; i < SZ; i++) {
        add_prev(next_arr[i], i);
    }

    memset(used, 0, sizeof(used));
    ll ans = 1;

    for (int i = 0; i < SZ; i++) {
        if (!used[i]) {
            /* Find root of cycle */
            int root = i;
            while (!used[root]) {
                used[root] = 1;
                root = next_arr[root];
            }

            /* Root is false: all subtrees independent */
            ll f_val = 1;
            for (int j = 0; j < prev_count[root]; j++) {
                Result res = helper_iterative(prevs[root][j], root);
                f_val = (f_val * ((res.f + res.t) % MOD)) % MOD;
            }

            /* Root is true: next[root] must be false */
            ll t_val;
            if (next_arr[root] == root) {
                t_val = 0;
            } else {
                Result r1 = helper_iterative(next_arr[root], root);
                Result r2 = helper_iterative(root, next_arr[root]);
                t_val = (r1.f * r2.t) % MOD;
            }

            ans = (ans * ((f_val + t_val) % MOD)) % MOD;
        }
    }

    printf("%lld\n", ans);

    for (int i = 0; i < SZ; i++) {
        if (prevs[i]) free(prevs[i]);
    }
    return 0;
}
